packages/contracts-bridge/contracts/test/BridgeRouter.t.sol

```diff
@@ -297,6 +716,266 @@ contract BridgeRouterTest is BridgeTest {
+        );
+        localToken.mint(address(bridgeRouter), tokenAmount);
+        vm.expectEmit(true, true, false, true, address(localToken));
+        emit Transfer(address(bridgeRouter), recipient, tokenAmount);
+        vm.expectEmit(true, true, true, true, address(bridgeRouter));
+        emit Receive(
+            12884901888012,
+            address(localToken),
+            recipient,
+            address(0),
+            tokenAmount
+        );
+        bridgeRouter.exposed_giveTokens(
+            origin,
+            nonce,
+            tokenId,
+            action,
+            recipient
+        );
+        assertEq(localToken.balanceOf(recipient), tokenAmount);
+    }
+
+    function test_giveTokensRemoteExistingRepresentationSucceeds() public {
+        uint32 origin = remoteDomain;
+        uint32 nonce = 12;
+        address recipient = address(0xBEEF);
+        uint256 tokenAmount = 100;
+        bytes32 tokenDetailsHash = "adsf";
+        address token = remoteTokenLocalAddress;
+        bytes memory action = abi.encodePacked(
+            BridgeMessage.Types.Transfer,
+            recipient.addressToBytes32(),
+            tokenAmount,
+            tokenDetailsHash
+        );
+        bytes memory tokenId = abi.encodePacked(
+            localDomain,
+            token.addressToBytes32()
+        );
+        vm.expectEmit(true, true, false, true, token);
+        // It mints new representations
+        emit Transfer(address(0), recipient, tokenAmount);
+        vm.expectEmit(true, true, true, true);
+        emit Receive(12884901888012, token, recipient, address(0), tokenAmount);
+        bridgeRouter.exposed_giveTokens(
+            origin,
+            nonce,
+            tokenId,
+            action,
+            recipient
+        );
+        assertEq(remoteToken.balanceOf(recipient), tokenAmount);
+        assertEq(BridgeToken(token).detailsHash(), tokenDetailsHash);
+    }
+
+    function test_giveTokensRemoteNewRepresentationSucceeds() public {
+        uint32 origin = remoteDomain;
+        uint32 nonce = 12;
+        address recipient = address(0xBEEF);
+        uint256 tokenAmount = 100;
+        bytes32 tokenDetailsHash = "adsf";
+        bytes32 token = "remote token addr";
+        bytes memory action = abi.encodePacked(
+            BridgeMessage.Types.Transfer,
+            recipient.addressToBytes32(),
+            tokenAmount,
+            tokenDetailsHash
+        );
+        bytes memory tokenId = abi.encodePacked(remoteDomain, token);
+        // As the token has no representation on the local domain
+        // bridgeRouter will ask TokenRegistry to deploy a new BridgeToken representation
+        // The address of the deployment is determenistic because it uses CREATE
+        address tokenRepresentation = computeCreateAddress(
+            address(tokenRegistry),
+            // we use nonce = 2, because this is the second contract deployed by
+            // tokenRegistry
+            2
```

Comment:
use `vm.getNonce` to avoid breaking during forks, etc

Reply:
🤯